### ISSUE TITLE
Dynamically loading an external webpack bundled ngModel using loadChildren

### ISSUE THREAD

**Issue by factoidforrest** [TheLarkInn/ama#11]
[This stack overflow ](https://stackoverflow.com/questions/47018621/dynamically-loading-an-external-webpack-bundled-ngmodel-as-a-route-handler)post is a good summation of the issue I hope.  Please let me know if you'd like to try to run the project and I will create a public repo.

We are trying to create "microservices on the client" by breaking apart apps into multiple projects and loading the chunks at runtime.  If I figure this out, I intend to create a blog post showing the community how to do this.  Could be a game changer for large teams working on huge front end apps.  

Help me Sean Kenobi, you're my only hope.  Also, our company is a financial contributor to both angular and webpack, for what that's worth.  ;)

**Comment by TheLarkInn** [comment]
Hi there! So I did spend some time looking at the C1 architecture. So is it possible? Likely, you would be leveraging webpack's Externals feature webpack.js.org/configuration/externals

How specifically in the context of Angular? That I'm not 100% sure about. 

My specific take is very opinionated and goes against the idea of Microservices on the web. But then again everything is tradeoffs! 

For one, I'm a very strong believer (maybe a bit biased but also because of the impact I've seen it has on web performance) in a single compilation. (aka webpack bundles everything at once, all sources are locally available in a monorepo, or in some way are pulled down at build time to be available, and nothing dynamic). The wins are: 

* Huge performance wins
* Ability to apply complex and performant caching strategies
* Fragile, if there is a problem in one piece of your app (some schools of thought would say _don't_ ship the rest but instead let a single point of failure represent an early warning detector for your whole user experience) 

And the cons: 

* Not antifragile, aka the whole thing about microservices that everyone raves about
* heavy duplication across all the dynamically loaded code

**Comment by factoidforrest** [comment]
Yeah Angular is really not cooperating with this idea.  That's been our struggle.  Not that it's impossible, but that angular breaks in all sorts of ways.  

SystemJS, which preserves module names even when bundling, works perfectly.  It's designed out of the box to have multiple independent bundles play nice at runtime.  Maybe someday webpack or the tools that follow it will have an option to preserve module names. 

The scary thing is that SystemJS + Angular may lose support soon.  And getting AOT working with SystemJS, especially with this dynamic setup, could be difficult or not feasible.  With Angular going AOT always on in v5, we are worried about the SystemJS approach, although it is working extremely well for now.  

I am leaning towards checking our chunked modules into npm(artifactory) and pulling them in at webpack build time, but the monolithic approach is not universally popular here.

**Comment by TheLarkInn** [comment]
@light24bulbs Honestly, I'd be extremely interested in this use case so I'm all for you trying this!

**Comment by factoidforrest** [comment]
Hey, we have made some serious progress towards this in the last month.  You might find it interesting.

We realized we could use DLL plugin to share everything we need between builds.  The only issue was that module names would still collide from separate bundles, so we just hooked into the webpack compilation process and turned them into UUIDs so they wouldn't collide.  (@vikikamath figured that one out)  

It's kind of amazing how nothing broke when we started customizing the module keys. 

What do you think?  


Hooking into the webpack compilation was the trickiest part, actually, and it still isn't perfect.  All the modules that DLLReferencePlugin injects seem to have normal integer IDs still, which is an issue.  Just trying to figure out how to fix that right now.

@rvadev

Here's a snippet from our plugin that handles the names 


```javascript
plugin.prototype.nameModule = function (module) {
  const base_dir_regex = new RegExp(this.options.baseDir)  //strip the dependency path down to local directories
  let name = null;
  const path = module.resource;
  console.log('visiting module ', path)
  console.log((`<<< ` + path).magenta);
  //name = uuidv4();
  if (path){
    name = uuidv4();
    // name = path.split(base_dir_regex)[1]
    // if (this.options.hash){ //maybe just add a hash for things not from node_modules to prevent application collisions
    //     let hash = md5File.sync(path);
    //     name = name + `#` + hash;
    // }
  }
  console.log((`>>>>: ` + name).cyan);
  return name;
}

plugin.prototype.apply = function (compiler) { //can also take a callback if we need to do async
    const self = this;
    console.log(`>>>> start renaming plugin using config ${inspect(this.options)}`.blue);
    compiler.plugin("compilation", function(compilation) {
        compilation.plugin("before-module-ids", function(modules) {
            console.log("++++ prefix: " + self.src_prefix);
            // console.log(modules);
            modules.forEach(module => {
                module.id = self.nameModule(module)
            })
            console.log("***** module ids: ********");
            // callback();
        })
    });
}
```

**Comment by TheLarkInn** [comment]
Do you think this is something you could create a proof of concept in a non-angular app? Like very simple minimal JS?

**Comment by factoidforrest** [comment]
Yeah, without a doubt.  You're interested in running it?  It has nothing to do with angular, really.

Angular has just been a huge pain in the butt actually, but we managed to trick AOT into working with this idea by having fake modules that get replaced by the ones from our dynamic bundles.

**Comment by factoidforrest** [comment]
If you know a better lifecycle hook than `before-module-ids` or have any comments about the way we are hooking in it would be awesome.  That hook isn't even in the documentation I don't think. https://github.com/webpack/docs/wiki/plugins
